### ISSUE TITLE
Better implementation of HTTP request generating and response parsing

### ISSUE THREAD

**Issue by CameronDiver** [resin-io-playground/narwhal#3]

**Comment by imrehg** [comment]
@CameronDiver I was looking into the current implementation because of the issues I've run into with #16 

Here's what I've found so far, why chunked encoding might fail currently:

1) the logic of "stop if your read couldn't fill the available buffer" is false in the case of observed docker response

https://github.com/resin-io-playground/narwhal/blob/1a7eba6b1ae67038d9d5fcfe489350c07a37b32d/src/httpstream.rs#L27-L29

I managed to get cases, when the server returns less than the whole reply, and have to retry reading from the stream, to get the next section.

So far couldn't come up with a scenario, that works in all cases:

* sometime we finish reading early, while if we had read more, we'd gotten data
* just retrying to read doesn't work, because if there's no data to read, then we'll just hang
* there's no obvious line ending from the reply that could be used to tell if we need to read more
* the `Read` trait's [`read_to_end`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_end) or [`read_to_string`](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_string) that would be a possible way to go about normally, does not work either, it never encounters the required `EOF` when reading from the `TcpStream`

Looking at other people's similar comments here: https://users.rust-lang.org/t/checking-for-incoming-data-on-tcpstream/5814/21

> Reading/writing from/to the TcpStream is also very hackish. Both read and write are not guaranteed to read/write the whole message.

Will likely need to set things up differently:

* with [BufRead](https://doc.rust-lang.org/std/io/trait.BufRead.html) trait and `read_until` or  `read_line`
* with wrapping things in a `peek`-able socket (?)
* with wrapping things in [`std::sync::mpsc::Receiver`](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html)
* ... something else?

**Comment by CameronDiver** [comment]
@imrehg Yeah I fully agree that the network side of things is a bit hacky. It's slightly weird, as you mention, because things are not guaranteed. Assuming that docker always closes the socket properly, asynchronous reads and writes might be the best way forward. 

We either get data, an `EAGAIN` or a disconnect, which we can handly nicely. This was my thoughts anyway, I haven't gotten around to implementing or investigating it :)
